### PR TITLE
feat: library readiness — exceptions, decoupled output, public API

### LINKED ISSUE #2: Fix circular import pattern in base.py
## Problem

`base.py` uses a lazy import pattern to avoid circular dependencies:

```python
def _get_console() -> "Console":
    from vodoo.main import console
    return console
```

This creates tight coupling between `base.py` and `main.py`.

## Suggested Fix

Consider one of:
1. Inject the console as a parameter to functions that need it
2. Use a proper dependency injection pattern
3. Create a separate `output.py` module for console management

## Location
`src/vodoo/base.py:16-25`

### LINKED ISSUE #3: Refactor global mutable state for console configuration
## Problem

`main.py` uses global mutable state for console configuration:

```python
_console_config = {"no_color": False}
console = Console()

# Later reassigned via:
global console  # noqa: PLW0603
console = get_console()
```

This pattern makes testing harder and could cause issues with concurrent usage.

## Suggested Fix

Consider using Typer's context object or a proper configuration singleton pattern.

## Location
`src/vodoo/main.py:121-123` and `main.py:160`

### LINKED ISSUE #4: Add missing test fixture and conftest.py
## Problem

`test_generic.py` references a `client` fixture that is not defined:

```python
def test_create_read_update_delete(client: OdooClient) -> None:
```

The test will fail without a `conftest.py` providing the fixture.

## Suggested Fix

Create `tests/conftest.py` with:
1. A `client` fixture (either mocked or using test credentials)
2. Consider adding pytest markers for integration tests that require a live Odoo instance

## Location
`tests/test_generic.py:14`

### LINKED ISSUE #5: Improve exception handling specificity in CLI commands
## Problem

CLI commands catch all exceptions broadly:

```python
except Exception as e:
    console.print(f"[red]Error:[/red] {e}")
    raise typer.Exit(1) from e
```

This masks specific errors and makes debugging harder.

## Suggested Fix

Catch specific exceptions separately:
- `xmlrpc.client.Fault` - Odoo API errors
- `ValueError` - Validation errors  
- `ConnectionError` - Network issues
- `RuntimeError` - Authentication failures

Each can have a more helpful error message.

## Location
Multiple locations in `src/vodoo/main.py` (e.g., lines 202-204, 232-234, etc.)

### LINKED ISSUE #6: Division operator returns float for potentially integer fields
## Problem

The division operator in `parse_field_assignment` always returns a float:

```python
elif operator == "/=":
    if parsed_value == 0:
        msg = "Division by zero"
        raise ValueError(msg)
    parsed_value = current_value / parsed_value  # Always returns float
```

If the Odoo field is an integer field, writing a float value may cause issues.

## Suggested Fix

Consider using integer division (`//`) when both operands are integers, or document this behavior clearly.

```python
if isinstance(current_value, int) and isinstance(parsed_value, int):
    parsed_value = current_value // parsed_value
else:
    parsed_value = current_value / parsed_value
```

## Location
`src/vodoo/base.py:936-940`

### LINKED ISSUE #7: HTML to markdown converter doesn't preserve link URLs
## Problem

The `HTMLToMarkdown` parser in `_html_to_markdown` doesn't preserve link hrefs:

```python
elif tag == "a":
    self.result.append("[")
# ...
elif tag == "a":
    self.result.append("]")  # Missing (url) part
```

Links are converted to `[text]` instead of `[text](url)`.

## Suggested Fix

Capture the href attribute in `handle_starttag` and append it in `handle_endtag`:

```python
def handle_starttag(self, tag, attrs):
    if tag == "a":
        self.result.append("[")
        self.current_href = dict(attrs).get("href", "")

def handle_endtag(self, tag):
    if tag == "a":
        self.result.append(f"]({self.current_href})")
```

## Location
`src/vodoo/base.py:369-402`

### LINKED ISSUE #8: Remove unused 'model' parameter from display_record_detail
## Problem

The `model` parameter in `display_record_detail` is declared but never used:

```python
def display_record_detail(
    record: dict[str, Any],
    model: str,  # noqa: ARG001  <-- Marked as unused
    show_html: bool = False,
    record_type: str = "Record",
) -> None:
```

## Suggested Fix

Either:
1. Remove the parameter if it's not needed
2. Use it (e.g., to generate a URL or fetch field metadata)

Note: This may require updating callers in helpdesk.py, project.py, and project_project.py.

## Location
`src/vodoo/base.py:185-190`

### LINKED ISSUE #14: Bug: create -d does not convert markdown to HTML for description field
## Summary

The `create -d` flag on resource commands (e.g. `project-task create`, `helpdesk create`) wraps the description text in raw `<p>` tags without applying markdown→HTML conversion. In contrast, the `set` subcommand correctly auto-converts markdown to HTML for HTML fields like `description`.

## Steps to Reproduce

```bash
# This produces raw markdown wrapped in <p> tags — no conversion
uvx vodoo project-task create "Test task" --project 2 -d "# Heading

- item 1
- item 2"

# This correctly converts markdown to <h1>, <ul>, etc.
uvx vodoo project-task set TASK_ID "description=# Heading

- item 1
- item 2"
```

## Expected Behavior

`create -d` should apply the same markdown→HTML conversion as `set` does for HTML fields.

## Actual Behavior

`create -d` stores the raw markdown text inside a `<p>` tag:
```html
<p># Heading

- item 1
- item 2</p>
```

While `set` correctly produces:
```html
<h1>Heading</h1>
<ul>
<li>item 1</li>
<li>item 2</li>
</ul>
```

## Affected Commands

Likely all `create -d` paths: `project-task create`, `helpdesk create`, etc.

## Workaround

Create the record first without `-d` (or with a placeholder), then immediately call `set` to apply the description with markdown conversion.

### PR DESCRIPTION
**Pull request by julian-r**
## Summary

Makes vodoo usable as both a CLI (`uvx vodoo`) and a Python library (`from vodoo import OdooClient`). Adds full documentation site, CI pipeline, async client, and exception hierarchy.

---

### Exception hierarchy (`src/vodoo/exceptions.py`)
- Full hierarchy under `VodooError` mirroring Odoo server exceptions
- `transport_error_from_data()` maps `data["name"]` to specific types (`OdooAccessError`, `OdooValidationError`, etc.)
- Domain-specific exceptions: `RecordNotFoundError`, `ConfigurationError`, `FieldParsingError`

### Decoupled output (`src/vodoo/base.py`)
- Replaced `from vodoo.main import console` with `configure_output()` — library imports no longer load the CLI module
- `vodoo.main` is not in `sys.modules` when using vodoo as a library

### Public API (`src/vodoo/__init__.py`)
- Exports `OdooClient`, `OdooConfig`, `Markdown`, `HTML`, and all exceptions via `__all__`

### Content types (`src/vodoo/content.py`)
- `Markdown` and `HTML` str subclasses for explicit control over HTML field conversion
- `client.create()` and `client.write()` auto-convert `Markdown(...)` to HTML before sending
- Domain functions (e.g. `create_task`) wrap description in `Markdown()` by default
- Zero overhead for bare strings — no extra RPC calls

### Transport layer (`src/vodoo/transport.py`, `src/vodoo/client.py`)
- Replaced `OdooTransportError` with new exception hierarchy
- Server errors are mapped to specific exception classes automatically
- Dropped deprecated `OdooRPCError` re-export

### HTTPS warning (`src/vodoo/config.py`)
- `OdooConfig` warns when URL does not use HTTPS (warns, does not error — localhost still works)

### Async client (`src/vodoo/aio/`)
- Full async implementation of the vodoo client and all domain modules
- Async transport layer for JSON-2 and legacy JSON-RPC
- Modules: `client.py`, `transport.py`, `auth.py`, `base.py`, `crm.py`, `helpdesk.py`, `project.py`, `project_project.py`, `knowledge.py`, `generic.py`, `security.py`, `timer.py`

### MkDocs documentation (`docs/`, `mkdocs.yml`)
- 34-page documentation site using MkDocs Material
- Getting Started (installation, configuration, quick start)
- CLI Reference (all 8 subcommands with examples)
- Library Guide (Python usage, AI assistant integration)
- API Reference (auto-generated from docstrings via mkdocstrings)
- Development (contributing, architecture, integration tests, security)
- Versioned docs with mike — deploy per release with version switcher

### CI pipeline (`.github/workflows/ci.yml`)
- Lint + unit tests on PRs
- Integration tests on push to main

### Publishing (`.github/workflows/publish.yml`)
- Deploy docs alongside PyPI release
- Switch to `hatch-vcs` for git-tag-based versioning (single source of truth)

### CLI improvements
- Replaced 76 broad `except Exception` blocks with specific exception handling (`RecordNotFoundError`, `OdooAccessError`, `TransportError`, `VodooError`, etc.)
- Error messages now distinguish between access errors, not-found, transport errors, and unexpected errors

### Code quality
- Deduplicated sync/async modules: extracted `_generate_password()`, `_match_field_assignment()`, `_parse_raw_value()`, `_apply_operator()` as shared pure functions
- HTML-to-markdown converter now preserves link URLs (`[text](url)`)
- Division operator returns int when both operands are int and result is exact
- Removed unused `model` parameter from `display_record_detail()`
- `get_project()` uses default detail fields to avoid Odoo 19 `stage_id` permission error
- Integration test conftest injects env vars to prevent production `.env` credential leakage

### README
- Rewritten to position vodoo as both a library and CLI
- Python quick-start examples, exception hierarchy diagram, library usage patterns
- CLI examples condensed by domain

### Other
- `CLAUDE.md` updated with integration test commands, docs commands, versioning guidance, and `gh` CLI usage
- `pyproject.toml` — added `docs` extra with mkdocs dependencies, hatch-vcs config
- Removed `test_generic.py` (was a duplicate of integration test without fixtures)

---

### Tests

- **30 unit tests** (`tests/test_exceptions.py`) with `MockTransport` — no network needed
- **4 integration tests** (`TestExceptions` in `tests/integration/test_suite.py`) — verified against Odoo 18 and 19
- **Full async integration tests** (`tests/integration/test_async_suite.py`)
- **93 integration tests pass** against Odoo 19 community edition

### Issues

Closes #2 — circular import in `base.py` eliminated
Closes #3 — global console state replaced with `configure_output()`
Closes #5 — CLI exception handling uses specific exception types
Closes #6 — division operator returns int for exact integer results
Closes #7 — HTML-to-markdown converter preserves link URLs
Closes #8 — unused `model` parameter removed from `display_record_detail`
Closes #14 — `create -d` converts markdown to HTML via `Markdown` content type

Closes #4 — `test_generic.py` removed (was an integration test without fixtures); exception tests now use `MockTransport`

### CODE REVIEW

**Review by julian-r — COMMENTED** [review]
test

File: src/vodoo/exceptions.py [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+
+
+# -- Configuration / connection ------------------------------------------------
+
+
+class ConfigurationError(VodooError):
+    """Raised when the configuration is invalid or incomplete."""
+
+
+class InsecureURLError(ConfigurationError):
```
Comment (julian-r) [inline src/vodoo/exceptions.py]: **Dead code: `InsecureURLError` defined but never raised**

The config validator only emits a `UserWarning` for insecure URLs — it never raises this exception. Either use it or remove it.

**Review by julian-r — COMMENTED** [review]
test

File: src/vodoo/base.py [hunk 2]
```diff
@@ -1,35 +1,67 @@
...
+# ---------------------------------------------------------------------------
+# Output configuration
+# ---------------------------------------------------------------------------
+# The CLI layer (main.py) calls ``configure_output()`` to set the active
+# console and simple-output flag.  When Vodoo is used as a library these
+# defaults are used instead, so no import of main.py is needed.
 
-def _get_console() -> "Console":
-    """Get console instance from main module.
+_output_console: Console | None = None
```
Comment (julian-r) [inline src/vodoo/base.py]: **Design: Global mutable state makes library non-reentrant**

`_output_console` and `_output_simple` are module-level globals. If two callers configure different output modes, they stomp on each other. Also makes testing harder.

Consider `contextvars.ContextVar` or passing context explicitly.

**Review by julian-r — COMMENTED** [review]
test

File: src/vodoo/helpdesk.py [hunk 3]
```diff
@@ -40,6 +40,18 @@
 MODEL = "helpdesk.ticket"
```
Comment (julian-r) [inline src/vodoo/helpdesk.py]: **Suggestion: Domain module boilerplate**

Every domain module (helpdesk, crm, project, project_project, knowledge) follows an identical pattern: `MODEL` constant + ~15 one-liner delegates to `base.py`. That's 12 files (6 sync + 6 async) of ~90% boilerplate (~2000 lines total).

A registry or factory approach could collapse each to ~20 lines:
```python
helpdesk = DomainModule("helpdesk.ticket", tag_model="helpdesk.tag", default_fields=[...])
```

**Review by julian-r — COMMENTED** [review]
test

**Review by julian-r — COMMENTED** [review]
test

**Comment by julian-r** [comment]
**Re: Issue #2 (Sync/Async Transport Duplication)**

Documented the intentional duplication in the `aio/transport.py` module docstring (e4bd22b). The ~260 lines of mechanical duplication are a deliberate trade-off — alternatives (`unasync`, codegen, runtime indirection) add build or type-system complexity that outweighs the cost for a library this size.

**Comment by julian-r** [comment]
**Re: Issue #5 (Async Client Race Condition) — FIXED**

Added double-checked `asyncio.Lock` to `AsyncOdooClient._ensure_transport()` (c14fcfc). Without the lock, concurrent coroutines on a fresh client could both enter initialization, leaking the first transport. The fast path (transport already initialized) takes no lock.

**Comment by julian-r** [comment]
**Re: Issue #7 (Global Mutable State in base.py)**

Documented the rationale in `base.py` (1f0f29c). The module-level globals are set once at CLI startup and read many times. `contextvars` would add complexity for a scenario that doesn't exist — the Odoo server itself is the shared mutable state, so tests and CLI sessions are inherently sequential.

**Comment by julian-r** [comment]
**Re: Issue #8 (Type Looseness in Domain Modules) — FIXED**

Replaced `Any` with correct types across all 10 domain modules (23363e0):
- `output_dir: Any` → `Path | None`
- `file_path: Any` → `Path | str`
- `-> list[Any]` → `-> list[Path]`

These now match the `base.py` signatures they delegate to.

**Comment by julian-r** [comment]
**Re: Issue #9 (Config Loading Stops at First Found) — FIXED**

Flipped the config file lookup order to local-overrides-global (d7c24f6):

1. `.vodoo.env` (project-specific)
2. `.env` (project generic)
3. `~/.config/vodoo/config.env` (global fallback)

This matches the convention used by git, npm, pip, and direnv. Still first-found-wins (no merging) — if you place a `.vodoo.env` in your project, it's the full config for that context.

**Comment by julian-r** [comment]
**Re: Issue #10 (Transport Detection Swallows Errors) — FIXED**

Changed `except Exception` to `except VodooError` in both sync and async `_detect_transport()` (df5d897). Verified against live Odoo 18 and 19 instances:

- Odoo 19: correctly detects JSON-2 (`is_json2=True`)
- Odoo 18: correctly falls back to Legacy (`is_json2=False`)
- Unreachable server: now raises `ConnectError` immediately instead of silently creating a doomed Legacy transport

Also investigated a lightweight GET probe to `/json/2` for detection, but both Odoo 18 and 19 return unhelpful status codes (404 vs 303) that aren't cleanly distinguishable without parsing response bodies. The `except VodooError` approach is simpler and sufficient.

**Comment by julian-r** [comment]
**Re: Issue #13 (Test Coverage for Pure Functions) — FIXED**

Added 67 unit tests across 3 new test files (8f0defe):

- `test_fields.py` (31 tests): `_match_field_assignment`, `_parse_raw_value`, `_apply_operator` — edge cases include multiline values, JSON prefix, division by zero, non-numeric operands
- `test_transport.py` (26 tests): `_build_json2_body`, `_parse_json2_response`, `_parse_name_search` — covers all method-specific branching, bare value parsing, malformed input
- `test_content.py` (10 tests): `process_values`, `Markdown`, `HTML` — type identity, conversion, pass-through

All run without an Odoo instance.

**Comment by julian-r** [comment]
## All 13 Review Issues — Resolved

| # | Issue | Resolution |
|---|---|---|
| 1 | Error handling boilerplate in main.py | **Fixed** — `_handle_errors()` context manager, −1342 lines |
| 2 | Sync/async transport duplication | **Documented** — intentional trade-off, alternatives add complexity |
| 3 | Domain module boilerplate | **Documented** — explicit delegates for discoverability |
| 4 | Client wraps transport unnecessarily | **Documented** — uniform `client.method()` API surface |
| 5 | Async client race condition | **Fixed** — `asyncio.Lock` with double-check pattern |
| 6 | base.py mixed concerns | **Fixed** — extracted `fields.py` |
| 7 | Global mutable output state | **Documented** — Odoo server is the real shared state |
| 8 | Type looseness (`Any` where `Path`) | **Fixed** — correct types in all 10 domain modules |
| 9 | Config loading precedence | **Fixed** — local overrides global (matches git/npm) |
| 10 | Transport detection swallows errors | **Fixed** — `except VodooError`, verified on Odoo 18 + 19 |
| 11 | Unused `InsecureURLError` | **Fixed** — removed dead code |
| 12 | Stale docstring parameter | **Fixed** — removed |
| 13 | Unit test coverage gaps | **Fixed** — 67 new tests (fields, transport, content) |
